src/table-procedure/src/create.rs

```diff
@@ -0,0 +1,397 @@
+use serde::{Deserialize, Serialize};
+use snafu::{OptionExt, ResultExt};
+use table::engine::{EngineContext, TableEngineProcedureRef, TableEngineRef, TableReference};
+use table::requests::CreateTableRequest;
+
+use crate::error::{
+    AccessCatalogSnafu, CatalogNotFoundSnafu, DeserializeProcedureSnafu, SchemaNotFoundSnafu,
+    SerializeProcedureSnafu,
+};
+
+/// Procedure to create a table.
+pub struct CreateTableProcedure {
+    data: CreateTableData,
+    catalog_manager: CatalogManagerRef,
+    table_engine: TableEngineRef,
+    engine_procedure: TableEngineProcedureRef,
+}
+
+#[async_trait]
+impl Procedure for CreateTableProcedure {
+    fn type_name(&self) -> &str {
+        Self::TYPE_NAME
+    }
+
+    async fn execute(&mut self, ctx: &Context) -> Result<Status> {
+        match self.data.state {
+            CreateTableState::Prepare => self.on_prepare(),
+            CreateTableState::EngineCreateTable => self.on_engine_create_table(ctx).await,
+            CreateTableState::RegisterCatalog => self.on_register_catalog().await,
+        }
+    }
+
+    fn dump(&self) -> Result<String> {
+        let json = serde_json::to_string(&self.data).context(SerializeProcedureSnafu)?;
+        Ok(json)
+    }
+
+    fn lock_key(&self) -> LockKey {
+        // We lock the whole table.
+        let table_name = self.data.table_ref().to_string();
+        LockKey::single(table_name)
+    }
+}
+
+impl CreateTableProcedure {
+    const TYPE_NAME: &str = "table-procedures::CreateTableProcedure";
+
+    /// Returns a new [CreateTableProcedure].
+    pub fn new(
+        request: CreateTableRequest,
+        catalog_manager: CatalogManagerRef,
+        table_engine: TableEngineRef,
+        engine_procedure: TableEngineProcedureRef,
+    ) -> CreateTableProcedure {
+        CreateTableProcedure {
+            data: CreateTableData {
+                state: CreateTableState::Prepare,
+                request,
+                sub_procedure_id: None,
+            },
+            catalog_manager,
+            table_engine,
+            engine_procedure,
+        }
+    }
+
+    /// Register the loader of this procedure to the `procedure_manager`.
+    ///
+    /// # Panics
+    /// Panics on error.
+    pub fn register_loader(
+        catalog_manager: CatalogManagerRef,
+        engine_procedure: TableEngineProcedureRef,
+        table_engine: TableEngineRef,
+        procedure_manager: ProcedureManagerRef,
+    ) {
+        procedure_manager
+            .register_loader(
+                Self::TYPE_NAME,
+                Box::new(move |data| {
+                    Self::from_json(
+                        data,
+                        catalog_manager.clone(),
+                        table_engine.clone(),
+                        engine_procedure.clone(),
+                    )
+                    .map(|p| Box::new(p) as _)
+                }),
+            )
+            .unwrap()
+    }
+
+    /// Recover the procedure from json.
+    fn from_json(
```

Comment:
Is it better to uplift deserialization method to the procedure trait? There's serialization method `dump` in that trait, which should be paired with some deserialization to stay consistent with each other I think.

Reply:
Different procedures require different components (e.g. catalog manager, table engine) so it's hard to define a general deserialize method.